lib/ex_unit/lib/ex_unit/assertions.ex

```diff
@@ -485,15 +524,29 @@ defmodule ExUnit.Assertions do
             timeout
           )
         end
+      end
 
     quote do
       timeout = unquote(timeout)
 
+      {:pattern, [], [left]} = unquote(left_pattern)
+      pattern = Pattern.new(left, unquote(pins), unquote(pattern_vars))
+
       {received, unquote(vars)} =
         receive do
           unquote(pattern) -> {received, unquote(vars)}
         after
-          timeout -> flunk(unquote(failure_message))
+          timeout ->
+            case unquote(failure_message) do
+              {:flunk, msg} ->
+                flunk(msg)
+
+              {:pattern, messages, msg} ->
+                assert false,
+                  right: {:ex_unit_mailbox_contents, messages},
```

Comment:
We shuold just pass `messages` instead of a special tuple and have a key in `ExUnit.Pattern` that says if it is meant to match a single element or multiple.